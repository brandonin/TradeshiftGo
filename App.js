import React from 'react';
import { NativeModules, LayoutAnimation, Image, TextInput, StyleSheet, Text, View } from 'react-native';
const { UIManager } = NativeModules;
UIManager.setLayoutAnimationEnabledExperimental &&
  UIManager.setLayoutAnimationEnabledExperimental(true);

export default class App extends React.Component {
  state = {
    inputs: {
      first: "20",
      second: "30",
      third: "20",
    },
    triangleTypes: "",
  }

  componentDidMount = () => {
    this.determineTriangleTypeAndSides();
  }


  handleChange = async (key, value) => {
    let inputs = {...this.state.inputs};
    inputs[key] = value;
    await this.setState({inputs});
    this.determineTriangleTypeAndSides();
  }

  determineTriangleTypeAndSides = () => {
    const {first, second, third} = this.state.inputs;
    let hypotenuse = -Infinity;
    let hypotenuseKey;
    for (const [key, value] of Object.entries(this.state.inputs)) {
      if (value > hypotenuse) {
        hypotenuse = value;
        hypotenuseKey = key;
      }
    }
    let [side1, side2] = assignSides(this.state.inputs);
    let triangleTypes;

    // triangle by sides
    if (hypotenuse === side1 && side1 === side2) triangleTypes = "equilateral";
    else if (hypotenuse === side1 || hypotenuse === side2 || side1 === side2) triangleTypes = "isosceles";
    else triangleTypes = 'scalene';

    if (this.inputIsNonZero()) {
      this.setState({triangleTypes, hypotenuse, side1, side2});
    }
    function assignSides(object) {
      for (const [key, value] of Object.entries(object)) {
        if (key === hypotenuseKey) {
          continue;
        }
        if (!side1) {
          side1 = value || 0;
        } else {
          side2 = value || 0;
        }
      }
      return [side1, side2];
    }
  }

  inputIsNonZero = () => {
    return Object.values(this.state.inputs).indexOf('') < 0 &&
      Object.values(this.state.inputs).indexOf('0') < 0;
  }

  dynamicStyling = () => {
    LayoutAnimation.easeInEaseOut();
    return {
      borderLeftWidth: parseInt(this.state.side1),
      borderRightWidth: parseInt(this.state.side2),
      borderBottomWidth: parseInt(this.state.hypotenuse),
    }
  }

  render() {
    const tsLogo = {
        uri: 'https://tradeshift.com/wp-content/themes/Tradeshift/img/brand/logo-black.png',
    };
    const { triangleTypes } = this.state;
    return (
      <View style={styles.container}>
        <View style={styles.topBanner}>
          <Image source={tsLogo} style={{width: 332, height: 40}}/>
        </View>
        <View style={styles.textContainer}>
          <View style={{ flexDirection:'row' }}>
            <Text>Length 1: </Text>
            <TextInput keyboardType="numeric" style={styles.input} value={this.state.inputs.first} placeholder="Length 1" onChangeText={(value) => this.handleChange('first', value)}></TextInput>
          </View>
          <View style={{ flexDirection:'row' }}>
            <Text>Length 2: </Text>
            <TextInput keyboardType="numeric" style={styles.input} value={this.state.inputs.second} placeholder="Length 2" onChangeText={(value) => this.handleChange('second', value)}></TextInput>
          </View>
          <View style={{ flexDirection:'row' }}>
            <Text>Length 3: </Text>
            <TextInput keyboardType="numeric" style={styles.input} value={this.state.inputs.third} placeholder="Length 3" onChangeText={(value) => this.handleChange('third', value)}></TextInput>
          </View>
          <Text>Type of Triangle: {triangleTypes}</Text>
          <View style = {[styles.triangle, this.dynamicStyling()]} />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 40,
  },
  textContainer: {
    paddingLeft: 15,
    paddingRight: 15,
  },
  topBanner: {
    alignItems: 'center',
    borderBottomWidth: 3,
    borderBottomColor: '#00b0ff',
    paddingTop: 10,
    paddingBottom: 10,
    width: '100%',
  },
  h1: {
    fontSize: 16,
    fontWeight: '900',
    paddingBottom: 15,
    paddingTop: 15,
    textAlign: 'center',
  },
  li: {
    paddingBottom: 15,
    paddingLeft: 25,
    paddingTop: 15,
  },
  triangle: {
    width: 0,
    height: 0,
    borderLeftColor: "transparent",
    borderRightColor: "transparent",
    borderBottomColor: 'red',
  },
  input: {
    backgroundColor: "rgba(180, 180, 180, 0.5)",
    textAlign: "center",
  }
});
