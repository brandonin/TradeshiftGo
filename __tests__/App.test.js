import React from 'react';
import App from '../App';
import { mount, shallow } from 'enzyme';
import {LayoutAnimation} from 'react-native';
jest.mock('LayoutAnimation');

let wrapper;
let state;
let spy;
describe('App Component', () => {
    beforeEach(() => {
        wrapper = shallow(<App />);
        state = wrapper.state();
    });

    it('should renders correctly without breaking', () => {
        expect(wrapper).toBeTruthy();
    });
    it('should renders three TextInputs', () => {
        expect(wrapper.find("TextInput")).toHaveLength(3);
    });

    it('should matche snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('Initialization', () => {
        it('should call determineTriangleTypeAndSides', () => {
            spy = jest.spyOn(wrapper.instance(), 'determineTriangleTypeAndSides')
            wrapper.instance().componentDidMount()
            expect(spy).toHaveBeenCalled();
        });
    });

    describe('Method determineTriangleTypeAndSides', () => {
        beforeEach(() => {
            wrapper.instance().determineTriangleTypeAndSides();
        });

        it('should call inputIsNonZero', () => {
            spy = jest.spyOn(wrapper.instance(), 'inputIsNonZero')
            wrapper.instance().determineTriangleTypeAndSides();
            expect(spy).toHaveBeenCalled();
        });

        describe('calculates type of triangle', () => {
            it('should determine an equilateral has all equal sides', async () => {
                await wrapper.setState({
                    inputs: {
                        first: '20',
                        second: '20',
                        third: '20',
                    }
                })
                await wrapper.instance().determineTriangleTypeAndSides();
                state = wrapper.state();
                expect(state.triangleTypes).toBe('equilateral');
            });

            it('should determine an isosceles if two sides are equal', async () => {
                await wrapper.setState({
                    inputs: {
                        first: '20',
                        second: '30',
                        third: '20',
                    }
                })
                await wrapper.instance().determineTriangleTypeAndSides();
                state = wrapper.state();
                expect(state.triangleTypes).toBe('isosceles');
            });

            it('should determine an scalene if no sides are equal', async () => {
                await wrapper.setState({
                    inputs: {
                        first: '20',
                        second: '30',
                        third: '10',
                    }
                })
                await wrapper.instance().determineTriangleTypeAndSides();
                state = wrapper.state();
                expect(state.triangleTypes).toBe('scalene');
            });
        })

        describe('updates state', () => {
            it('should calculate hypotenuse', () => {
                expect(state.hypotenuse).toEqual('30');
            });

            it('should make hyptoenuse the longest side', () => {
                expect(parseInt(state.hypotenuse)).toBeGreaterThanOrEqual(parseInt(state.side1));
                expect(parseInt(state.hypotenuse)).toBeGreaterThanOrEqual(parseInt(state.side2));
            });

            it('should calculate triangleTypes', () => {
                expect(state.triangleTypes).toEqual('isosceles');
            });

            it('should calculate side1', () => {
                expect(state.side1).toEqual('20');
            });

            it('should calculate side2', () => {
                expect(state.side2).toEqual('20');
            });
        });
    });

    describe('Method inputIsNonZero', () => {
        it('should return false if input is zero', () => {
            let inputs = {...state.inputs}
            inputs['side1'] = "0";
            wrapper.setState({inputs});
            const doesNotHaveZero = wrapper.instance().inputIsNonZero();
            expect(doesNotHaveZero).toBeFalsy();
        });

        it('should return false if inputs are non-zero', () => {
            let inputs = {...state.inputs}
            inputs['side1'] = "30";
            wrapper.setState({inputs});
            const doesNotHaveZero = wrapper.instance().inputIsNonZero();
            expect(doesNotHaveZero).toBeTruthy();
        });
    });

    describe('method dynamicStyling', () => {
        beforeEach(() => {
            wrapper.instance().dynamicStyling();
        });
        it('should have animations', () => {
            expect(LayoutAnimation.easeInEaseOut).toHaveBeenCalled();
        });

        it('should parse strings to numbers', () => {
            Object.values(wrapper.instance().dynamicStyling()).forEach((value) => {
                expect(typeof value).toBe('number');
            })
        })
    });

    describe('handleChange', () => {
        beforeEach(async () => {
            spy = jest.spyOn(wrapper.instance(), 'determineTriangleTypeAndSides');
            await wrapper.instance().handleChange('first', '40');
            state = wrapper.state();
        });

        it('should only change one state at a time', () => {
            expect(state.inputs.first).toBe('40');
            expect(state.inputs.second).toBe('30');
            expect(state.inputs.third).toBe('20');
        });
        it('should call determineTriangleTypeAndSides', () => {
            expect(spy).toHaveBeenCalled();
        });
    })
});
