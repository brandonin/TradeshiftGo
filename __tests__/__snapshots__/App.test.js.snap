// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapsot 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "paddingTop": 40,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#00b0ff",
        "borderBottomWidth": 3,
        "paddingBottom": 10,
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <Component
      source={
        Object {
          "uri": "https://tradeshift.com/wp-content/themes/Tradeshift/img/brand/logo-black.png",
        }
      }
      style={
        Object {
          "height": 40,
          "width": 332,
        }
      }
    />
  </Component>
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 1: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 1"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 2: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 2"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="30"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 3: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 3"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component>
      Type of Triangle: 
      isosceles
    </Component>
    <Component
      style={
        Array [
          Object {
            "borderBottomColor": "red",
            "borderLeftColor": "transparent",
            "borderRightColor": "transparent",
            "height": 0,
            "width": 0,
          },
          Object {
            "borderBottomWidth": 30,
            "borderLeftWidth": 20,
            "borderRightWidth": 20,
          },
        ]
      }
    />
  </Component>
</Component>
`;

exports[`App Component renders correctly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "paddingTop": 40,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#00b0ff",
        "borderBottomWidth": 3,
        "paddingBottom": 10,
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <Component
      source={
        Object {
          "uri": "https://tradeshift.com/wp-content/themes/Tradeshift/img/brand/logo-black.png",
        }
      }
      style={
        Object {
          "height": 40,
          "width": 332,
        }
      }
    />
  </Component>
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 1: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 1"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 2: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 2"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="30"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        underlineColorAndroid="transparent"
      >
        Length 3: 
      </TextInput>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 3"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component>
      Type of Triangle: 
      isosceles
    </Component>
    <Component
      style={
        Array [
          Object {
            "borderBottomColor": "red",
            "borderLeftColor": "transparent",
            "borderRightColor": "transparent",
            "height": 0,
            "width": 0,
          },
          Object {
            "borderBottomWidth": 30,
            "borderLeftWidth": 20,
            "borderRightWidth": 20,
          },
        ]
      }
    />
  </Component>
</Component>
`;

exports[`App Component should matche snapshot 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "paddingTop": 40,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#00b0ff",
        "borderBottomWidth": 3,
        "paddingBottom": 10,
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <Component
      source={
        Object {
          "uri": "https://tradeshift.com/wp-content/themes/Tradeshift/img/brand/logo-black.png",
        }
      }
      style={
        Object {
          "height": 40,
          "width": 332,
        }
      }
    />
  </Component>
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Component>
        Length 1: 
      </Component>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 1"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Component>
        Length 2: 
      </Component>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 2"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="30"
      />
    </Component>
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Component>
        Length 3: 
      </Component>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="Length 3"
        style={
          Object {
            "backgroundColor": "rgba(180, 180, 180, 0.5)",
            "textAlign": "center",
          }
        }
        underlineColorAndroid="transparent"
        value="20"
      />
    </Component>
    <Component>
      Type of Triangle: 
      isosceles
    </Component>
    <Component
      style={
        Array [
          Object {
            "borderBottomColor": "red",
            "borderLeftColor": "transparent",
            "borderRightColor": "transparent",
            "height": 0,
            "width": 0,
          },
          Object {
            "borderBottomWidth": 30,
            "borderLeftWidth": 20,
            "borderRightWidth": 20,
          },
        ]
      }
    />
  </Component>
</Component>
`;
